src/base/bittorrent/sessionimpl.cpp

```diff
@@ -5867,20 +5883,50 @@ void SessionImpl::handleTorrentConflictAlert(const lt::torrent_conflict_alert *a
 
 void SessionImpl::processTrackerStatuses()
 {
+    if (m_updatedTrackerEntries.isEmpty())
+        return;
+
     for (auto it = m_updatedTrackerEntries.cbegin(); it != m_updatedTrackerEntries.cend(); ++it)
     {
-        auto torrent = static_cast<TorrentImpl *>(it.key());
-        const QSet<QString> &updatedTrackers = it.value();
+        invokeAsync([this, torrentHandle = it.key(), updatedTrackers = it.value()]() mutable
+        {
+            try
+            {
+                std::vector<lt::announce_entry> nativeTrackers = torrentHandle.trackers();
+                invoke([this, torrentHandle, nativeTrackers = std::move(nativeTrackers)
+                        , updatedTrackers = std::move(updatedTrackers)]
+                {
+                    TorrentImpl *torrent = m_torrents.value(torrentHandle.info_hash());
+                    if (!torrent)
+                        return;
 
-        for (const QString &trackerURL : updatedTrackers)
-            torrent->invalidateTrackerEntry(trackerURL);
-    }
+                    QHash<QString, TrackerEntry> updatedTrackerEntries;
+                    updatedTrackerEntries.reserve(updatedTrackers.size());
+                    for (const lt::announce_entry &announceEntry : nativeTrackers)
+                    {
+                        const auto updatedTrackersIter = updatedTrackers.find(announceEntry.url);
+                        if (updatedTrackersIter == updatedTrackers.end())
+                            continue;
 
-    if (!m_updatedTrackerEntries.isEmpty())
-    {
-        emit trackerEntriesUpdated(m_updatedTrackerEntries);
-        m_updatedTrackerEntries.clear();
+                        const QMap<TrackerEntry::Endpoint, int> &updateInfo = updatedTrackersIter.value();
+                        const TrackerEntry trackerEntry = torrent->updateTrackerEntry(announceEntry, updateInfo);
+#if (QT_VERSION < QT_VERSION_CHECK(6, 0, 0))
+                        updatedTrackerEntries[trackerEntry.url] = std::move(trackerEntry);
+#else
+                        updatedTrackerEntries.emplace(trackerEntry.url, std::move(trackerEntry));
```

Comment:
> the correct action is to declare `trackerEntry` as non-const just so we can use the move constructor/assignment here

I think Yes. Thank you!

Reply:
Just a C++ rant: It feels like a hack. Ideally we want to have `const` here to denote that we don't want to change the value and have the compiler catch usage errors. But we are "forced" to drop that semantic in order to have std::move work as intended.

Reply:
Yes, C++ does not allow us to be "both smart and beautiful at the same time" in such cases.
In fact, not all such constants are really required, they often exist for better readability, debugging, etc., i.e. to store an intermediate value that otherwise could simply be omitted:
```c++
updatedTrackerEntries.emplace(trackerEntry.url, torrent->updateTrackerEntry(announceEntry, updateInfo));
```